Kentor.AuthServices/Configuration/ServiceCertificateCollection.cs

```diff
@@ -1,58 +1,72 @@
+using System.Collections.ObjectModel;
+using System.Reflection;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Xml;
 
 namespace Kentor.AuthServices.Configuration
 {
     /// <summary>
-    /// Config collection of ServiceCertificateElements.
+    /// Certificates used by the service provider for signing or decryption.
     /// </summary>
-    public class ServiceCertificateCollection : ConfigurationElementCollection, IEnumerable<ServiceCertificateElement>
+    public class ServiceCertificateCollection: Collection<ServiceCertificate>
     {
         /// <summary>
-        /// Create new element of right type.
+        /// Add a certificate to the collection with default status use and
+        /// metadata behaviour.
         /// </summary>
-        /// <returns>ServiceCertificateElement</returns>
-        protected override ConfigurationElement CreateNewElement()
+        /// <param name="certificate">Certificate to add.</param>
+        public void Add(X509Certificate2 certificate)
         {
-            return new ServiceCertificateElement();
+            if (certificate == null) throw new ArgumentNullException(nameof(certificate));
+            InsertItem(base.Count, new ServiceCertificate
+            {
+                Certificate = certificate
+            });
         }
 
         /// <summary>
-        /// Get the name of an element.
+        /// Add to the collection at the specified position.
         /// </summary>
-        /// <param name="element">ServiceCertificateElement</param>
-        /// <returns>element.Name</returns>
-        protected override object GetElementKey(ConfigurationElement element)
+        /// <param name="index">Position index.</param>
+        /// <param name="item">Service certificate to add.</param>
+        protected override void InsertItem(int index, ServiceCertificate item)
         {
-            return Guid.NewGuid().ToString();
-        }
+            if (item == null) throw new ArgumentNullException(nameof(item));
 
-        /// <summary>
-        /// Get a strongly typed enumerator.
-        /// </summary>
-        /// <returns>Strongly typed enumerator.</returns>
-        public new IEnumerator<ServiceCertificateElement> GetEnumerator()
-        {
-            return base.GetEnumerator().AsGeneric<ServiceCertificateElement>();
+            if (!item.Certificate.HasPrivateKey)
+            {
+                throw new ArgumentException(@"Provided certificate is not valid for encryption/decryption because it does not contain a private key.");
+            }
+
+            if (item.Use == CertificateUse.Encryption || item.Use == CertificateUse.Both)
+            {
+                if (!CertificateWorksForDecryption(item.Certificate))
+                {
+                    throw new ArgumentException(@"Provided certificate is not valid for encryption/decryption. It can only be used for other purposes such as signing.");
```

Comment:
Is it possible to be a bit more clear on what to do? Maybe add something like "if you only want to use the certificate for signing, you should add a Usage=Signing config".